File: scratch/deepmind/part_two/find-rotation-point.ts
function findRotationPoint(xs: Array<string>): number {
  // Find the rotation point in the vector.
  let beg = 0;
  let end = xs.length - 1;

  while (beg != end) {
    let mid = beg + Math.floor((end - beg) / 2);

    if (beg === mid) {
      return xs[beg] < xs[end] ? beg : end;
    }

    if (xs[end] <= xs[mid]) {
      beg = mid;
      end = end;
    } else {
      beg = beg;
      end = mid;
    }
  }

  return beg;
}

// Tests
let desc;
let actual;
let expected;

desc = "small array one";
actual = findRotationPoint(["cape", "cake"]);
expected = 1;
assertEquals(actual, expected, desc);

desc = "small array two";
actual = findRotationPoint(["cake", "cape"]);
expected = 0;
assertEquals(actual, expected, desc);

desc = "medium array";
actual = findRotationPoint(["grape", "orange", "plum", "radish", "apple"]);
expected = 4;
assertEquals(actual, expected, desc);

desc = "large array";
actual = findRotationPoint([
  "ptolemaic",
  "retrograde",
  "supplant",
  "undulate",
  "xenoepist",
  "asymptote",
  "babka",
  "banoffee",
  "engender",
  "karpatka",
  "othellolagkage"
]);
expected = 5;
assertEquals(actual, expected, desc);

function assertEquals(a, b, desc) {
  if (a === b) {
    console.log(`${desc} ... PASS`);
  } else {
    console.log(`${desc} ... FAIL: ${a} != ${b}`);
  }
}

File: users/wpcarro/scratch/deepmind/part_one/find-rotation-point.py
import unittest
from math import floor


def midpoint(a, b):
    return a + floor((b - a) / 2)


def do_find_rotation_point(a, b, xs):
    i = midpoint(a, b)
    count = b - a + 1

    if count == 2:
        if xs[a] > xs[b]:
            return b
        else:
            return -1

    if i in {a, b}:
        return i

    if xs[a] < xs[i]:
        return do_find_rotation_point(i, b, xs)
    else:
        return do_find_rotation_point(a, i, xs)


def find_rotation_point(xs):
    return do_find_rotation_point(0, len(xs) - 1, xs)


# Tests
class Test(unittest.TestCase):
    def test_small_list(self):
        actual = find_rotation_point(['cape', 'cake'])
        expected = 1
        self.assertEqual(actual, expected)

    def test_medium_list(self):
        actual = find_rotation_point(
            ['grape', 'orange', 'plum', 'radish', 'apple'])
        expected = 4
        self.assertEqual(actual, expected)

    def test_large_list(self):
        actual = find_rotation_point([
            'ptolemaic', 'retrograde', 'supplant', 'undulate', 'xenoepist',
            'asymptote', 'babka', 'banoffee', 'engender', 'karpatka',
            'othellolagkage'
        ])
        expected = 5
        self.assertEqual(actual, expected)


unittest.main(verbosity=2)
